### PR TITLE
[SPARK-35214][SQL] OptimizeSkewedJoin support ShuffledHashJoinExec

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add `ShuffledHashJoin` pattern check in `OptimizeSkewedJoin` so that we can optimize it.

### Why are the changes needed?

Currently, we have already supported all type of join through hint that make it easy to choose the join implementation.

We would choose `ShuffledHashJoin` if one table is not big but over the broadcast threshold. It's better that we can support optimize it in `OptimizeSkewedJoin`.

### Does this PR introduce _any_ user-facing change?

Probably yes, the execute plan in AQE mode may be changed.

### How was this patch tested?

Improve exists test in `AdaptiveQueryExecSuite`

### CODE REVIEW

**Comment by ulysses-you** [comment]
retest this please

**Comment by ulysses-you** [comment]
Do you have time to take a look ? Thank you. @maropu @cloud-fan @JkSelf @maryannxue

**Comment by maropu** [comment]
Also, cc: @c21

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledJoin.scala [hunk 1]
```diff
@@ -19,15 +19,23 @@ package org.apache.spark.sql.execution.joins
...
 import org.apache.spark.sql.catalyst.plans.{ExistenceJoin, FullOuter, InnerLike, LeftExistence, LeftOuter, RightOuter}
-import org.apache.spark.sql.catalyst.plans.physical.{Distribution, HashClusteredDistribution, Partitioning, PartitioningCollection, UnknownPartitioning}
+import org.apache.spark.sql.catalyst.plans.physical.{Distribution, HashClusteredDistribution, Partitioning, PartitioningCollection, UnknownPartitioning, UnspecifiedDistribution}
 
 /**
  * Holds common logic for join operators by shuffling two child relations
  * using the join keys.
  */
 trait ShuffledJoin extends BaseJoinExec {
+  def isSkewJoin: Boolean
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledJoin.scala:29]: Please update the description in SQLConf? https://github.com/apache/spark/blob/38ef4771d447f6135382ee2767b3f32b96cb1b0e/sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala#L541-L548

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala [hunk 2]
```diff
@@ -52,16 +51,6 @@ case class SortMergeJoinExec(
 
   override def stringArgs: Iterator[Any] = super.stringArgs.toSeq.dropRight(1).iterator
 
-  override def requiredChildDistribution: Seq[Distribution] = {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala:55]: Could you move `nodeName`, too?
```
  override def nodeName: String = {
    if (isSkewJoin) super.nodeName + "(skew=true)" else super.nodeName
  }
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 3]
```diff
@@ -157,98 +157,121 @@ object OptimizeSkewedJoin extends CustomShuffleReaderRule {
...
-          if (skewSpecs.isDefined) {
-            logDebug(s"Left side partition $partitionIndex " +
-              s"(${FileUtils.byteCountToDisplaySize(leftActualSize)}) is skewed, " +
-              s"split it into ${skewSpecs.get.length} parts.")
-            numSkewedLeft += 1
-          }
-          skewSpecs.getOrElse(Seq(leftPartSpec))
-        } else {
-          Seq(leftPartSpec)
+  private def getOptimizedChildren(
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:160]: nit: How about `getOptimizedChildren` -> `tryToOptimizedChildren`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 4]
```diff
@@ -157,98 +157,121 @@ object OptimizeSkewedJoin extends CustomShuffleReaderRule {
...
       }
+    }
+    logDebug(s"number of skewed partitions: left $numSkewedLeft, right $numSkewedRight")
+    if (numSkewedLeft > 0 || numSkewedRight > 0) {
+      Some((CustomShuffleReaderExec(left.shuffleStage, leftSidePartitions.toSeq),
+        CustomShuffleReaderExec(right.shuffleStage, rightSidePartitions.toSeq)))
+    } else {
+      None
+    }
+  }
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:242]: nit: It looks a unnecessary change.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 5]
```diff
@@ -685,64 +691,82 @@ class AdaptiveQueryExecSuite
...
-
-        // skewed right outer join optimization
-        val (_, rightAdaptivePlan) = runAdaptiveAndVerifyResult(
-          "SELECT * FROM skewData1 right outer join skewData2 ON key1 = key2")
-        val rightSmj = findTopLevelSortMergeJoin(rightAdaptivePlan)
-        checkSkewJoin(rightSmj, 0, 1)
+          // skewed inner join optimization
+          val (_, innerAdaptivePlan) = runAdaptiveAndVerifyResult(
+            s"SELECT /*+ $joinHint(skewData1) */ * FROM skewData1 " +
+              s"JOIN skewData2 ON key1 = key2")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:740]: nit: remove `s`.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 6]
```diff
@@ -685,64 +691,82 @@ class AdaptiveQueryExecSuite
...
+            findTopLevelSortMergeJoin(innerAdaptivePlan)
+          } else {
+            findTopLevelShuffledHashJoin(innerAdaptivePlan)
+          }
+          checkSkewJoin(inner, 2, 1)
+
+          // skewed left outer join optimization
+          val (_, leftAdaptivePlan) = runAdaptiveAndVerifyResult(
+            s"SELECT /*+ $joinHint(skewData2) */ * FROM skewData1 " +
+              s"LEFT OUTER JOIN skewData2 ON key1 = key2")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:751]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 7]
```diff
@@ -685,64 +691,82 @@ class AdaptiveQueryExecSuite
...
+            findTopLevelSortMergeJoin(leftAdaptivePlan)
+          } else {
+            findTopLevelShuffledHashJoin(leftAdaptivePlan)
+          }
+          checkSkewJoin(leftSmj, 2, 0)
+
+          // skewed right outer join optimization
+          val (_, rightAdaptivePlan) = runAdaptiveAndVerifyResult(
+            s"SELECT /*+ $joinHint(skewData1) */ * FROM skewData1 " +
+              s"RIGHT OUTER JOIN skewData2 ON key1 = key2")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:762]: ditto

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala (hunk 4 above)
Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:242]: Do you mean the `None` ?  Not sure I see your point.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 8]
```diff
@@ -157,98 +157,121 @@ object OptimizeSkewedJoin extends CustomShuffleReaderRule {
...
-          if (skewSpecs.isDefined) {
-            logDebug(s"Left side partition $partitionIndex " +
-              s"(${FileUtils.byteCountToDisplaySize(leftActualSize)}) is skewed, " +
-              s"split it into ${skewSpecs.get.length} parts.")
-            numSkewedLeft += 1
-          }
-          skewSpecs.getOrElse(Seq(leftPartSpec))
-        } else {
-          Seq(leftPartSpec)
+  private def tryToOptimizedChildren(
```
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:160]: nit: the `origin smj` in above comment needs to be updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala [hunk 9]
```diff
@@ -231,11 +231,12 @@ object EnsureRequirements extends Rule[SparkPlan] {
...
-      case ShuffledHashJoinExec(leftKeys, rightKeys, joinType, buildSide, condition, left, right) =>
+      case ShuffledHashJoinExec(
+        leftKeys, rightKeys, joinType, buildSide, condition, left, right, isSkew) =>
         val (reorderedLeftKeys, reorderedRightKeys) =
           reorderJoinKeys(leftKeys, rightKeys, left.outputPartitioning, right.outputPartitioning)
         ShuffledHashJoinExec(reorderedLeftKeys, reorderedRightKeys, joinType, buildSide, condition,
-          left, right)
+          left, right, isSkew)
 
       case SortMergeJoinExec(leftKeys, rightKeys, joinType, condition, left, right, isPartial) =>
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:241]: totally nit: what about renaming `isPartial` to `isSkew` here?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala (hunk 4 above)
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:242]: Ah, okay. In the last commit, I found the unnecessary change removing the blank line.

**Review by c21 — COMMENTED** [review]
With restriction of `OptimizeSkewedJoin.canSplitLeft()` (and similarly `OptimizeSkewedJoin.canSplitRight()`), here we only can handle skew join for stream side of shuffled hash join, right? I think this PR itself is still valuable.

But the main concerning point from discussion of https://github.com/apache/spark/pull/32210 is to handle skew join of build side. Not sure if there's a fundamental reason that whether we can support skew join handling of build side.

**Comment by c21** [comment]
btw just some random thought here. E.g. for `LEFT OUTER JOIN` (build right side. stream left side), and we found left side is skewed and split the skewed partition `L1` into 3 smaller partitions (just an example here). Would the corresponding partition on right side `R1` needs to be built hash map 3 times for `L1`'s 3 smaller partitions? If that's true, it sounds to me that it may potentially introduce more OOM on build side, as tasks are sharing executor's off-heap memory to build hash maps?

**Comment by ulysses-you** [comment]
@c21  thanks for the input.

Yes,  we can not optimize skew with all type of joins's build side in AQE. But at least, currently we can handle the skew inner like join with both stream and build side.

> If that's true, it sounds to me that it may potentially introduce more OOM on build side, as tasks are sharing executor's off-heap memory to build hash maps

Yes, it's a side effect for `OptimizeSkewedJoin`, but smj's advantage is it could spill. IMO, if user specify the shuffled hash join to do execution that means they know the benefit and issue of it. And in the other hand, we can easily increase the menroy but hard to make skew join fast.

**Comment by c21** [comment]
> Yes, it's a side effect for OptimizeSkewedJoin, but smj's advantage is it could spill. IMO, if user specify the shuffled hash join to do execution that means they know the benefit and issue of it. And in the other hand, we can easily increase the memory but hard to make skew join fast. So this optimization can be the extra choice for user.

I agree this adds extra choice for user given current status of thing. But in the long-term, we would like to work towards enabling shuffled hash join by default (i.e. `spark.sql.join.preferSortMergeJoin`=false). This seems to me add more [risk](https://github.com/apache/spark/pull/32328#issuecomment-826601325) to the long term direction. So I think we should be more cautious with it and have more discussion. cc @cloud-fan.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala (hunk 8 above)
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:160]: `tryOptimizeJoinChildren`?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala [hunk 10]
```diff
@@ -685,64 +691,82 @@ class AdaptiveQueryExecSuite
...
-          .createOrReplaceTempView("skewData1")
-        spark
-          .range(0, 1000, 1, 10)
-          .select(
-            when('id < 250, 249)
-              .otherwise('id).as("key2"),
-            'id as "value2")
-          .createOrReplaceTempView("skewData2")
+    Seq("SHUFFLE_MERGE", "SHUFFLE_HASH").foreach { joinHint =>
+      val isSMJ = joinHint == "SHUFFLE_MERGE"
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/adaptive/AdaptiveQueryExecSuite.scala:695]: nit: to reduce more duplicated code
```
def getJoinNode(plan: LogicalPlan) = if (joinHint == "SHUFFLE_MERGE") {
  findTopLevelSortMergeJoin(plan)
} else {
  findTopLevelShuffledHashJoin(plan)
}
```

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
The overhead of skew join handling is to replicate the non-skew side, and the overhead is larger in SHJ because we need to build the hash relation. But the overhead of skew should be more significant than this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 11]
```diff
@@ -157,98 +157,121 @@ object OptimizeSkewedJoin extends CustomShuffleReaderRule {
...
+  }
 
-      logDebug(s"number of skewed partitions: left $numSkewedLeft, right $numSkewedRight")
-      if (numSkewedLeft > 0 || numSkewedRight > 0) {
-        val newLeft = CustomShuffleReaderExec(left.shuffleStage, leftSidePartitions.toSeq)
-        val newRight = CustomShuffleReaderExec(right.shuffleStage, rightSidePartitions.toSeq)
+  def optimizeSkewJoin(plan: SparkPlan): SparkPlan = plan.transformUp {
+    case smj @ SortMergeJoinExec(_, _, joinType, _,
+        s1 @ SortExec(_, _, ShuffleStage(left: ShuffleStageInfo), _),
+        s2 @ SortExec(_, _, ShuffleStage(right: ShuffleStageInfo), _), _)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:253]: nit: let's make sure we don't repeatedly optimize skew join. The pattern match should make sure `isSkewJoin` is false.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 12]
```diff
@@ -157,98 +157,121 @@ object OptimizeSkewedJoin extends CustomShuffleReaderRule {
...
+        val (newLeft, newRight) = newChildren.get
         smj.copy(
           left = s1.copy(child = newLeft), right = s2.copy(child = newRight), isSkewJoin = true)
       } else {
         smj
       }
+
+    case shj @ ShuffledHashJoinExec(_, _, joinType, _, _,
+        ShuffleStage(left: ShuffleStageInfo),
+        ShuffleStage(right: ShuffleStageInfo), _)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:266]: ditto

**Comment by c21** [comment]
> I don't think this PR makes OOM much more likely to happen. It does add more tasks that need to build a hash relation, but the parallelism on the executor JVM is not increased.

@cloud-fan - yes you are right. I miss to think for the parallelism part. Then I think this PR is good to go and no risk. Sorry for chiming in, @ulysses-you .

**Comment by ulysses-you** [comment]
@c21 ah, thank you, I'm gold to have some discussion, and also missed consider parallelism part : )  @cloud-fan thank you, addressed all comment.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala [hunk 13]
```diff
@@ -148,7 +148,7 @@ object OptimizeSkewedJoin extends CustomShuffleReaderRule {
   /*
    * This method aim to optimize the skewed join with the following steps:
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:149]: immature thoughts on this optimization:

`OptimizeSkewedJoin` try to make sizes of shuffle partitions more even. 

For bhj, should we change the *optimization goal of build side* to making sure build side fit in memory to avoid OOM or fallback to smj.

Is it possible that the build side is skewed according current criterions, but the shuffle partitions all fit in memory?

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:149]: In current implementaion, we have already supported optimize skew build side with inner join. For other join type we cann't optimize it due to the semantics.

IMO It's better to consider OOM at shj itself instead of fallback to smj which might make things more complicated. Actually, we might change smj to shj at `reOptimize` if we disable the `preferSortMerge`, so it's confused to change join strategy again.

Comment (zhengruifeng) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:149]: Sorry my comment was misleading, the `fallback to smj` there means https://github.com/apache/spark/pull/32210

My thought is that we may take potential OOM into account here,
let build side A inner join stream side B, for a build partition A_0:

- if A_0 is skewed but is less than a OOM threshold, we may not split it;
- if A_0 is not skewed but greater than that threshold, we may still need to split it;

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeSkewedJoin.scala:149]: My hunch is when the build side can be potentially OOM-ed, it should already be considered as skewed. So after AQE skew handling, some of potentially OOM-ed build side (inner join only) can be avoided.

However, for queries with other join types, queries not having shuffle before join, and queries with run-time hash map being significantly larger than partition size, we should have run-time fallback mechanism in shuffled hash join itself. This PR and #32210 should be good to have and orthogonal to each other.

**Review by maropu — APPROVED** [review]
This PR itself looks fine.

**Comment by maropu** [comment]
Thanks! Merged to master.

**Comment by ulysses-you** [comment]
thank you all !
